File: src/com/gmail/marcosav2010/cipher/CipheredCommunicator.java
package com.gmail.marcosav2010.cipher;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.IOException;
import java.security.GeneralSecurityException;
import java.security.Key;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import com.gmail.marcosav2010.common.Utils;
import com.gmail.marcosav2010.communicator.BaseCommunicator;
import com.gmail.marcosav2010.communicator.Communicator;
import com.gmail.marcosav2010.logger.Logger;
import com.gmail.marcosav2010.main.Main;
import com.gmail.marcosav2010.tasker.Task;
import com.gmail.marcosav2010.tasker.TaskOwner;

/**
 * This is a communicator that encrypts byte arrays with a random symmetric key under @SessionCipher
 * asymmetric encryption, it also handles other message parameters.
 * 
 * @author Marcos
 *
 */
public class CipheredCommunicator extends Communicator {

	private static final String CIPHER_SYMMETRIC_ALGORITHM = "AES/GCM/PKCS5Padding";

	private static final int IV_SIZE = 96;
	private static final int TAG_LENGTH = 128;
	public static final String AES_KEY_ALGORITHM = "AES";
	public static final int AES_KEY_SIZE = 256;

	private static final int AES_KEY_BYTES = AES_KEY_SIZE / Byte.SIZE;

	private static final int LENGTH_BYTES = Integer.BYTES;

	private static final int AES_DATA_SIZE = AES_KEY_BYTES + IV_SIZE/* + LENGTH_BYTES */; // Para AES256: 32 + 96 + 4 = 132
	private static final int RSA_MSG_SIZE = SessionCipher.RSA_KEY_SIZE / Byte.SIZE;

	private final SessionCipher sessionCipher;

	private final TaskOwner taskOwner;

	private Task writeTask;
	private WritePool writePool;

	public CipheredCommunicator(BaseCommunicator baseCommunicator, SessionCipher sessionCipher, TaskOwner taskOwner) {
		setIn(new BufferedInputStream(baseCommunicator.getIn()));
		setOut(new BufferedOutputStream(baseCommunicator.getOut()));

		this.sessionCipher = sessionCipher;
		this.taskOwner = taskOwner;
		start();
	}

	private void start() {
		writeTask = Main.getInstance().getTasker().run(taskOwner, writePool = new WritePool());
	}

	public synchronized EncryptedMessage read() throws IOException {
		// Leer tamaño del mensaje cifrado en complemento
		
		byte[] lengthBytes = getIn().readNBytes(LENGTH_BYTES);
		if (lengthBytes.length <= 0)
			return null;
		
		int length = ~Utils.bytesToInt(lengthBytes);
		if (length < 0)
			return null;

		// Leer parte RSA donde esta la llave AES
		byte[] encryptedAESData = getIn().readNBytes(RSA_MSG_SIZE);

		// Leer mensaje encriptado con AES de longitud dada antes
		byte[] encryptedData = getIn().readNBytes(length);

		return new EncryptedMessage(encryptedAESData, encryptedData);
	}

	public EncryptedMessage read(TaskOwner taskOwner, long timeout, TimeUnit unit)
			throws IOException, InterruptedException, ExecutionException, TimeoutException {
		return taskOwner.getExecutorService().submit(() -> read()).get(timeout, unit);
	}

	public void decrypt(EncryptedMessage encryptedMessage, Consumer<byte[]> onDecrypt) {
		if (encryptedMessage == null)
			return;

		Main.getInstance().getTasker().run(taskOwner, () -> {
			try {
				byte[] simmetricKeyBytes = sessionCipher.decode(encryptedMessage.getEncryptedSimmetricKeyBytes());

				var simmetricKeyData = deconstructAESData(simmetricKeyBytes);

				byte[] decryptedSimmetricKey = simmetricKeyData[0];
				byte[] decryptedSimmetricKeyIV = simmetricKeyData[1];

				SecretKey symmetricKey = getSecretKey(decryptedSimmetricKey, AES_KEY_ALGORITHM);
				GCMParameterSpec gcmPS = getGCMParameterSpec(decryptedSimmetricKeyIV);

				onDecrypt.accept(getCipher(Cipher.DECRYPT_MODE, symmetricKey, gcmPS).doFinal(encryptedMessage.getEncryptedData()));

			} catch (GeneralSecurityException | InterruptedException e) {
				Logger.log(e);
			}
		});
	}

	private class WritePool implements Runnable {

		private BlockingQueue<EncryptedMessage> queue = new LinkedBlockingQueue<>();
		private AtomicBoolean write = new AtomicBoolean(true);

		public void run() {
			while (write.get())
				try {
					write(queue.take());
				} catch (IOException e) {
					Logger.log(e);
				} catch (InterruptedException e) {
				}
		}

		public void queue(EncryptedMessage msg) {
			queue.offer(msg);
		}

		private void write(EncryptedMessage msg) throws IOException {
			getOut().write(msg.byteLength());
			getOut().write(msg.getEncryptedSimmetricKeyBytes());
			getOut().write(msg.getEncryptedData());
			getOut().flush();
		}

		public void cancel() {
			write.set(false);
			queue.clear();
		}
	}

	@Override
	public void write(byte[] bytes) {
		try {
			writePool.queue(encrypt(bytes));

		} catch (GeneralSecurityException | InterruptedException ex) {
			Logger.log(ex);
		}
	}

	public EncryptedMessage encrypt(byte[] bytes) throws GeneralSecurityException, InterruptedException {
		GCMParameterSpec gcmPS = generateGCMParameterSpec();
		byte[] iv = gcmPS.getIV();
		SecretKey symmetricKey = generateAESSecretKey();

		byte[] encryptedData = getCipher(Cipher.ENCRYPT_MODE, symmetricKey, gcmPS).doFinal(bytes);

		byte[] aesDataToRSAEncrypt = constructAESData(symmetricKey.getEncoded(), iv);
		byte[] encryptedAESData = sessionCipher.encode(aesDataToRSAEncrypt);

		return new EncryptedMessage(encryptedAESData, encryptedData);
	}

	public void beforeClose() {
		writePool.cancel();
		writeTask.cancelNow();
	}

	private static SecretKey generateAESSecretKey() throws NoSuchAlgorithmException {
		KeyGenerator symmetricKeyGen = KeyGenerator.getInstance(AES_KEY_ALGORITHM);
		symmetricKeyGen.init(AES_KEY_SIZE, new SecureRandom());
		return symmetricKeyGen.generateKey();
	}

	private static GCMParameterSpec generateGCMParameterSpec() {
		byte[] iv = new byte[IV_SIZE];
		new SecureRandom().nextBytes(iv);
		return new GCMParameterSpec(TAG_LENGTH, iv);
	}

	private static Cipher getCipher(int mode, Key key, GCMParameterSpec gcmParameterSpec) throws GeneralSecurityException {
		Cipher c = Cipher.getInstance(CIPHER_SYMMETRIC_ALGORITHM);
		c.init(mode, key, gcmParameterSpec);
		return c;
	}

	private static GCMParameterSpec getGCMParameterSpec(byte[] iv) {
		return new GCMParameterSpec(TAG_LENGTH, iv);
	}

	private static SecretKey getSecretKey(byte[] key, String algorithm) {
		return new SecretKeySpec(key, 0, key.length, algorithm);
	}

	private static byte[] constructAESData(byte[] simKey, byte[] iv/* , byte[] length */) {
		byte[] out = new byte[AES_DATA_SIZE];

		System.arraycopy(simKey, 0, out, 0, AES_KEY_BYTES);
		System.arraycopy(iv, 0, out, AES_KEY_BYTES, IV_SIZE);
		// System.arraycopy(length, 0, out, AES_KEY_BYTES + IV_SIZE, LENGTH_BYTES);

		return out;
	}

	private static byte[][] deconstructAESData(byte[] b) {
		byte[] keyBytes = new byte[AES_KEY_BYTES];
		byte[] ivBytes = new byte[IV_SIZE];
		// byte[] lengthBytes = new byte[LENGTH_BYTES];

		System.arraycopy(b, 0, keyBytes, 0, AES_KEY_BYTES);
		System.arraycopy(b, AES_KEY_BYTES, ivBytes, 0, IV_SIZE);
		// System.arraycopy(b, AES_KEY_BYTES + IV_SIZE, lengthBytes, 0, LENGTH_BYTES);

		return new byte[][] { keyBytes, ivBytes/* , lengthBytes */ };
	}

	@Override
	public void close() throws IOException {
		beforeClose();
		if (getIn() != null)
			getIn().close();

		if (getOut() != null)
			getOut().close();
	}

	@Override
	public void closeQuietly() {
		beforeClose();
		try {
			close();
		} catch (IOException ex) {
		}
	}
}

File: core/src/main/java/com/gmail/marcosav2010/cipher/CipheredCommunicator.kt
package com.gmail.marcosav2010.cipher

import com.gmail.marcosav2010.common.Utils
import com.gmail.marcosav2010.communicator.BaseCommunicator
import com.gmail.marcosav2010.communicator.Communicator
import com.gmail.marcosav2010.logger.ILog
import com.gmail.marcosav2010.logger.Log
import com.gmail.marcosav2010.tasker.Task
import com.gmail.marcosav2010.tasker.TaskOwner
import com.gmail.marcosav2010.tasker.Tasker.run
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.IOException
import java.security.GeneralSecurityException
import java.security.Key
import java.security.SecureRandom
import java.util.concurrent.BlockingQueue
import java.util.concurrent.Callable
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * This is a communicator that encrypts byte arrays with a random symmetric key
 * under @SessionCipher asymmetric encryption, it also handles other message
 * parameters.
 *
 * @author Marcos
 */
class CipheredCommunicator(baseCommunicator: BaseCommunicator,
                           private val sessionCipher: SessionCipher,
                           private val taskOwner: TaskOwner) :
        Communicator() {

    private val log: ILog = Log(taskOwner, "CC")

    private val writeTask: Task
    private val writePool = WritePool()

    init {
        input = BufferedInputStream(baseCommunicator.input!!)
        output = BufferedOutputStream(baseCommunicator.output!!)

        writeTask = run(taskOwner, writePool)
    }

    @Synchronized
    fun read(): EncryptedMessage? {
        // Leer tamaño del mensaje cifrado en complemento
        val lengthBytes = input!!.readNBytes(LENGTH_BYTES)
        if (lengthBytes.isEmpty()) return null

        val length = Utils.bytesToInt(lengthBytes).inv()
        if (length < 0) return null

        // Leer parte RSA donde esta la llave AES
        val encryptedAESData = input!!.readNBytes(RSA_MSG_SIZE)

        // Leer mensaje encriptado con AES de longitud dada antes
        val encryptedData = input!!.readNBytes(length)

        return EncryptedMessage(encryptedAESData, encryptedData)
    }

    fun read(taskOwner: TaskOwner, timeout: Long, unit: TimeUnit): EncryptedMessage =
            taskOwner.executorService.submit(Callable { read() })[timeout, unit]!!

    fun decrypt(encryptedMessage: EncryptedMessage?, onDecrypt: (ByteArray) -> Unit) {
        if (encryptedMessage == null) return

        run(taskOwner) {
            try {
                val symmetricKeyBytes = sessionCipher.decode(encryptedMessage.encryptedSymmetricKeyBytes)
                val symmetricKeyData = deconstructAESData(symmetricKeyBytes)
                val decryptedSymmetricKey = symmetricKeyData[0]
                val decryptedSymmetricKeyIV = symmetricKeyData[1]

                val symmetricKey = getSecretKey(decryptedSymmetricKey)
                val gcmPS = getGCMParameterSpec(decryptedSymmetricKeyIV)

                onDecrypt.invoke(getCipher(Cipher.DECRYPT_MODE, symmetricKey, gcmPS)
                        .doFinal(encryptedMessage.encryptedData))
            } catch (e: GeneralSecurityException) {
                log.log(e)
            } catch (e: InterruptedException) {
                log.log(e)
            }
        }
    }

    override fun write(bytes: ByteArray) {
        try {
            writePool.queue(encrypt(bytes))
        } catch (ex: GeneralSecurityException) {
            log.log(ex)
        } catch (ex: InterruptedException) {
            log.log(ex)
        }
    }

    private fun encrypt(bytes: ByteArray): EncryptedMessage {
        val gcmPS = generateGCMParameterSpec()
        val iv = gcmPS.iv
        val symmetricKey = generateAESSecretKey()
        val encryptedData = getCipher(Cipher.ENCRYPT_MODE, symmetricKey, gcmPS).doFinal(bytes)
        val aesDataToRSAEncrypt = constructAESData(symmetricKey.encoded, iv)
        val encryptedAESData = sessionCipher.encode(aesDataToRSAEncrypt)

        return EncryptedMessage(encryptedAESData, encryptedData)
    }

    private fun beforeClose() {
        writePool.cancel()
        writeTask.cancelNow()
    }

    override fun close() {
        beforeClose()

        input?.close()
        output?.close()
    }

    override fun closeQuietly() {
        beforeClose()
        try {
            close()
        } catch (ignored: IOException) {
        }
    }

    private inner class WritePool : () -> Unit {

        private val queue: BlockingQueue<EncryptedMessage> = LinkedBlockingQueue()
        private val write = AtomicBoolean(true)

        override fun invoke() {
            while (write.get()) try {
                write(queue.take())
            } catch (e: IOException) {
                log.log(e)
            } catch (ignored: InterruptedException) {
            }
        }

        fun queue(msg: EncryptedMessage) = queue.offer(msg)

        private fun write(msg: EncryptedMessage) {
            output!!.write(msg.byteLength)
            output!!.write(msg.encryptedSymmetricKeyBytes)
            output!!.write(msg.encryptedData)
            output!!.flush()
        }

        fun cancel() {
            write.set(false)
            queue.clear()
        }
    }

    companion object {
        const val AES_KEY_ALGORITHM = "AES"
        const val AES_KEY_SIZE = 256
        private const val CIPHER_SYMMETRIC_ALGORITHM = "AES/GCM/NoPadding" // JDK 14 does not recognize AES/GCM/PKCS5Padding
        private const val IV_SIZE = 96
        private const val TAG_LENGTH = 128
        private const val AES_KEY_BYTES = AES_KEY_SIZE / java.lang.Byte.SIZE
        private const val LENGTH_BYTES = Integer.BYTES
        private const val AES_DATA_SIZE = AES_KEY_BYTES + IV_SIZE /* + LENGTH_BYTES */ // For AES256: 32 + 96 + 4 = 132
        private const val RSA_MSG_SIZE = SessionCipher.RSA_KEY_SIZE / java.lang.Byte.SIZE

        private fun generateAESSecretKey(): SecretKey {
            val symmetricKeyGen = KeyGenerator.getInstance(AES_KEY_ALGORITHM)
            symmetricKeyGen.init(AES_KEY_SIZE, SecureRandom())
            return symmetricKeyGen.generateKey()
        }

        private fun generateGCMParameterSpec(): GCMParameterSpec {
            val iv = ByteArray(IV_SIZE)
            SecureRandom().nextBytes(iv)
            return GCMParameterSpec(TAG_LENGTH, iv)
        }

        private fun getCipher(mode: Int, key: Key, gcmParameterSpec: GCMParameterSpec): Cipher {
            val c = Cipher.getInstance(CIPHER_SYMMETRIC_ALGORITHM)
            c.init(mode, key, gcmParameterSpec)
            return c
        }

        private fun getGCMParameterSpec(iv: ByteArray) = GCMParameterSpec(TAG_LENGTH, iv)

        private fun getSecretKey(key: ByteArray) = SecretKeySpec(key, 0, key.size, AES_KEY_ALGORITHM)

        private fun constructAESData(simKey: ByteArray, iv: ByteArray /* , byte[] length */): ByteArray {
            val out = ByteArray(AES_DATA_SIZE)
            System.arraycopy(simKey, 0, out, 0, AES_KEY_BYTES)
            System.arraycopy(iv, 0, out, AES_KEY_BYTES, IV_SIZE)
            // System.arraycopy(length, 0, out, AES_KEY_BYTES + IV_SIZE, LENGTH_BYTES);
            return out
        }

        private fun deconstructAESData(b: ByteArray): Array<ByteArray> {
            val keyBytes = ByteArray(AES_KEY_BYTES)
            val ivBytes = ByteArray(IV_SIZE)
            // byte[] lengthBytes = new byte[LENGTH_BYTES];
            System.arraycopy(b, 0, keyBytes, 0, AES_KEY_BYTES)
            System.arraycopy(b, AES_KEY_BYTES, ivBytes, 0, IV_SIZE)
            // System.arraycopy(b, AES_KEY_BYTES + IV_SIZE, lengthBytes, 0, LENGTH_BYTES);
            return arrayOf(keyBytes, ivBytes /* , lengthBytes */)
        }
    }
}